sway-core/src/asm_generation/fuel/optimizations/symbolic_interpretation.rs

```diff
@@ -0,0 +1,232 @@
+    Full,
+    /// Reset non-virtual registers in addition to defs
+    NonVirtual,
+    /// Only the `def_registers` and `def_const_registers` are reset
+    Defs,
+}
+impl ResetKnown {
+    fn apply(&self, op: &Op, known_values: &mut KnownValues) {
+        match self {
+            ResetKnown::Full => {
+                known_values.values.clear();
+            }
+            ResetKnown::NonVirtual => {
+                Self::Defs.apply(op, known_values);
+                known_values
+                    .values
+                    .retain(|k, _| matches!(k, VirtualRegister::Virtual(_)));
+            }
+            ResetKnown::Defs => {
+                for d in op.def_registers() {
+                    known_values.clear_dependent_on(d);
+                    known_values.values.remove(d);
+                }
+                for d in op.def_const_registers() {
+                    known_values.clear_dependent_on(d);
+                    known_values.values.remove(d);
+                }
+            }
+        }
+    }
+}
+
+impl AbstractInstructionSet {
+    /// Symbolically interpret code and propagate known register values.
+    pub(crate) fn interpret_propagate(mut self) -> AbstractInstructionSet {
+        if self.ops.is_empty() {
+            return self;
+        }
+
+        // The set of labels that are jump targets
+        // todo: build proper control flow graph instead
+        let jump_target_labels: FxHashSet<Label> = self
+            .ops
+            .iter()
+            .filter_map(|op| match op.opcode {
+                Either::Right(
+                    ControlFlowOp::Jump(label)
+                    | ControlFlowOp::JumpIfNotZero(_, label)
+                    | ControlFlowOp::Call(label),
+                ) => Some(label),
+                _ => None,
+            })
+            .collect();
+
+        let mut known_values = KnownValues::default();
+
+        for op in &mut self.ops {
+            // Perform constant propagation on the instruction.
+            let mut uses_regs: Vec<_> = op.use_registers_mut().into_iter().collect();
+            for reg in uses_regs.iter_mut() {
+                // We only optimize over virtual registers here, constant registers shouldn't be replaced
+                if !reg.is_virtual() {
+                    continue;
+                }
+                if let Some(r) = known_values.resolve(reg).and_then(|r| r.register()) {
+                    **reg = r;
+                }
+            }
+
+            // Some instructions can be further simplified with the known values.
+            match &mut op.opcode {
+                // Conditional jumps can be simplified if we know the value of the register.
+                Either::Right(ControlFlowOp::JumpIfNotZero(reg, lab)) => {
+                    if let Some(con) = known_values.resolve(reg).and_then(|r| r.value()) {
+                        if con == 0 {
+                            op.opcode = Either::Left(VirtualOp::NOOP);
```

Comment:
Decrement `jump_target_label[lab]` here, and if it turns `0`, delete the key.

Reply:
That's true. This is already handled by doing multiple passes, but if done like this, we could get more done in a single pass.

Reply:
Done in da598c443